packages/eslint-plugin-next/src/rules/no-async-client-component.ts

```diff
@@ -0,0 +1,68 @@
+import { defineRule } from '../utils/define-rule'
+
+const url = 'https://nextjs.org/docs/messages/no-async-client-component'
+const description = 'Prevent client components from being async functions.'
+const message = `${description} See: ${url}`
+
+export = defineRule({
+  meta: {
+    docs: {
+      description,
+      recommended: true,
+      url,
+    },
+    type: 'problem',
+    schema: [],
+  },
+
+  create(context) {
+    return {
+      Program(node) {
+        let isClientComponent: boolean = false
+
+        for (const block of node.body) {
+          if (
+            block.type === 'ExpressionStatement' &&
+            block.expression.type === 'Literal' &&
+            block.expression.value === 'use client'
```

Comment:
The tricky thing about this being a linter rule is that, even if a file doesn't have `"use client"` on top, it can still be a Client Component (e.g. it can be imported by a file with `"use client"`). And also, a file with `"use client"` can still export non-component functions that are async. Not strongly against adding such a linter rule, but we can probably make it a warning instead of an error, and ensure that the exported function's name start with a capitalized character.

Reply:
Thank you for the review @shuding! I pushed the changes you suggested.

Agree this as a linter rule isn't a complete solution. In addition to the case you mentioned it also does not handle anonymous default exports or named exports. It is low-hanging fruit that should help prevent the page crashes people are reporting in https://github.com/vercel/next.js/issues/50898 and https://github.com/vercel/next.js/issues/50382, at least for ESLint users using Next's recommended config.

Linter rule aside, is it possible to throw an error at runtime for this scenario? That way it it avoids the page crash and provides some recourse to developers in the error message. As it stands right now there is no indication why the page crash happens

Reply:
Thanks! Yeah it sounds good to have this shipped first, at the same time we'll check how this can be improved over time.

Unfortunately we can't add this as a runtime check. We can't abort synchronous operations in that case.